File: app/src/core/model/round.ts
// Copyright (C) Call of Nil contributors
// SPDX-License-Identifier: AGPL-3.0-only

import * as commands from '@/commands';

export class RoundImpl implements Round {
  public readonly id: number;
  public readonly state: RoundState;

  private constructor(round: Round) {
    this.id = round.id;
    this.state = round.state;
  }

  public isIdle() {
    return this.state.kind === 'idle';
  }

  public isDone() {
    return this.state.kind === 'done';
  }

  public isWaiting() {
    return this.state.kind === 'waiting';
  }

  public isWaitingPlayer(id: PlayerId) {
    return this.isPlayerPending(id) || this.isPlayerReady(id);
  }

  public isPlayerPending(id: PlayerId) {
    return this.state.kind === 'waiting' && this.state.pending.includes(id);
  }

  public isPlayerReady(id: PlayerId) {
    return this.state.kind === 'waiting' && this.state.ready.includes(id);
  }

  public static create(round: Round) {
    if (round instanceof RoundImpl) {
      return round;
    }

    return new RoundImpl(round);
  }

  public static async load() {
    const round = await commands.getRound();
    return RoundImpl.create(round);
  }
}

File: crates/nil-core/src/world/cheat/round.rs
// Copyright (C) Call of Nil contributors
// SPDX-License-Identifier: AGPL-3.0-only

use crate::bail_if_cheats_are_not_allowed;
use crate::error::Result;
use crate::world::World;
use std::num::NonZeroU8;

pub fn skip_round(world: &mut World, amount: NonZeroU8) -> Result<()> {
  bail_if_cheats_are_not_allowed!(world);

  if !world.round.is_idle() {
    let amount = amount.get();
    for i in 1..=amount {
      world.dangerously_end_round(i == amount)?;
    }
  }

  Ok(())
}
